### PR TITLE
Add factdb peewee models

### PR DESCRIPTION
**Pull request by maxnoe**
This adds peewee models created by an automatic conversion script for our `factdata` database. These should be a lot nicer to use than writing raw sql. 

Probably I will rewrite the qla stuff to use this in the following days, but we should not wait for this to merge here.

### CODE REVIEW

**Comment by dneise** [comment]
How is this suposed to be used, can you add some form of documentation in the README perhaps or link this PR to an issue, which is presenting the problem this is going to solve?

**Comment by maxnoe** [comment]
This is introducing ORM models using peewee to remove the meed to write raw sql in python for queries to our fact database.

Instead you can now do stuff like:

```python
from fact.factdb import connect_database, RunInfo

connect_database()  # this uses the credentials module if no config is given

num_runs = RunInfo.select().count()
```

Or slightly more advanced:
```
num_crab_runs = (
    RunInfo
    .select()
    .join(Source, on=Source.fsourcekey == RunInfo.fsourcekey)
    .where(Source.fsourcename == 'Crab')
    .count()
)
````

Or if you want to have a list of dictionaries for night, run id and ontime for all data runs:

```
runs = list(
    RunInfo
    .select(RunInfo.fnight, RunInfo.frunid, RunInfo.fontime)
    .where(RunInfo.fruntypekey==1)
    .dicts()
)
```
Which could directly be converted to a `pandas.DataFrame`
```
df = pd.DataFrame(runs)
```

But this is really just repeating what you can doo with peewee from the peewee docs:
http://docs.peewee-orm.com/en/latest/peewee/quickstart.html

**Comment by maxnoe** [comment]
I added two convenience functions and two examples to the README

**Review by kbruegge — APPROVED** [review]
Looks good. Naive is weird. But I'm not an peewee expert. Whatever. I think this is useful.

**Comment by dneise** [comment]
This is what I wanted to understand, there are two ways I see to get the result of a DB request back as a pandas.DataFrame, which I assume everybody wants :-D

One: Let the peewee model interact with the Database, return a list of dicts, which the pandas. DataFrame constructor knows how to convert into a DataFrame. 
```
df = pd.DataFrame(list(
    RunInfo
    .select(RunInfo.fnight, RunInfo.frunid, RunInfo.fontime)
    .where(RunInfo.fruntypekey==1)
    .dicts()
))
```

Two: Let peewee generate the SQL for you, but let pandas interact with the Database. 
```
df = pd.read_sql(RunInfo
    .select(RunInfo.fnight, RunInfo.frunid, RunInfo.fontime)
    .where(RunInfo.fruntypekey==1).sql(),
    con=create_factdb_engine())
```

Not sure which one is "better", but I wanted to understand which is the way you would propose.

**Comment by maxnoe** [comment]
I get an error running the second one:
```
TypeError: not all arguments converted during string formatting
```
query.sql() returns a tuple of the query and the arguments which is the way pymysql expects a query when I run it.

What works for me is:
```
query, params = (
    RunInfo
    .select(RunInfo.fnight, RunInfo.frunid, RunInfo.fontime)
    .where(RunInfo.fruntypekey==1)
    .sql()
)
df = pd.read_sql(query, create_factdb_engine(), params=params)
```

I have no idea what is preferable. (Both use pymysql in the end i think)

**Comment by kbruegge** [comment]
Just from skimming it with one closed eye, I prefer the second version. Seems easier to read. No need for `dict` and `list` words.

**Comment by maxnoe** [comment]
I agree

**Comment by dneise** [comment]
Ah there is another thing which just came to my mind. 
What I found so intriguing when Max first showed me pandas Dataframes, was that they are so very well suited to avoid wrinting complicated SQL queries altogether. 

There is even this part in the pandas docu:
http://pandas.pydata.org/pandas-docs/stable/comparison_with_sql.html

So I personally always read the entire table (if I know by heart it is small enough) and try to make it somehow persistent, be it HDF5 or pickle or shelve or json ... 
and to all the joins, merges, and god knows what complicated operations I do locally ... 



In the end ... let's just merge this ... but I don't know how useful it might be in the end ... since this is now the 3rd method I see to get information out of the DB:

 * `pd.read_table("RunInfo")` followed by selections done locally
 * `pd.read_sql(peewee.make_sql_for_me())`
 * `peewee.get_a_list_of_dicts_for_me()`--> feed it into DataFrame ctor

I think this gets slowly into conflict with zen of python: 
"There should be one-- and preferably only one --obvious way to do it."

So by adding a new way of doing things, I think it is also our duty to give some sort of advice to fellow developers using `pyfact` in order to try to keep the number of ways people do things a little limited.

**Comment by dneise** [comment]
Anyway ... certainly you need this feature .. otherwise you wouldn't have gone through the trouble of implementing it ... so let's merge it

**Comment by maxnoe** [comment]
I changed the functions to use the pandas approach. It saved us some slocs!

File: fact/factdb/utils.py [hunk 1]
```diff
@@ -42,13 +38,8 @@ def get_ontime_by_source(runtype=None, config=None):
...
 
-    query = (
-        query
-        .group_by(Source.fsourcename)
-        .dicts()
-        .naive()  # put selected foreign key attributes directly into the objects
-    )
-    df = pd.DataFrame(list(query))
+    query, params = query.group_by(Source.fsourcename).sql()
+    df = pd.read_sql(query, engine or create_factdb_engine(), params=params)
```
Comment (dneise) [inline fact/factdb/utils.py:42]: wow ... this `engine or create_factdb_engine()` strongly reminds me of this `do() or die()` pattern known in php.
smells strange to me ... but looks awesome ...

Comment (maxnoe) [reply fact/factdb/utils.py:42]: This is actually a very common pattern in python for optional arguments

**Pull request merged**